In [2]:
from sklearn.datasets import make_regression
import numpy as np

In [4]:
X,Y=make_regression(n_samples=100,n_features=1, n_informative=1, n_targets=1,noise=20,random_state=13)

In [5]:
from sklearn.model_selection import train_test_split
xTrain,xTest,yTrain,yTest=train_test_split(X,Y,test_size=0.2,random_state=42)

In [6]:
from sklearn.linear_model import LinearRegression
lr=LinearRegression()
lr.fit(xTrain,yTrain)
y_pred=lr.predict(xTest)

In [7]:
from sklearn.metrics import r2_score
print("R2:",r2_score(yTest,y_pred))

R2: 0.6186926408914155


In [8]:
lr.coef_

array([28.16404017])

In [9]:
lr.intercept_

np.float64(-2.7130725669241036)

In [10]:
m=lr.coef_[0]
m

np.float64(28.16404017122941)

# **For finding b-intercept**

In [11]:
class GDRegressor():
  def __init__(self,learning_rate,epochs):
    self.b=0
    self.m=28.164
    self.lr=learning_rate
    self.epochs=epochs
  
  def fit(self,xTrain,yTrain):
    for i in range(self.epochs):
      loss_func=-2*np.sum(yTrain-self.m*xTrain.ravel()-self.b)
      self.b=self.b-(self.lr*loss_func)
    print(self.b)


In [12]:
gdr=GDRegressor(0.1,10)
gdr.fit(xTrain,yTrain)

1564493790867.8276


There is a very high diffrence between expected answer and what we actually got this is because our learnng rate is way too high:

In [13]:
gdr=GDRegressor(0.01,10)
gdr.fit(xTrain,yTrain)

-2.6966666813390243


Now as we can see it is closer to the expexted b intercept (-2.71) but still not equal now we increase the epochs:

In [14]:
gdr=GDRegressor(0.01,100)
gdr.fit(xTrain,yTrain)

-2.7130715877515823
